Firestore/core/src/firebase/firestore/util/hashing.h

```diff
@@ -0,0 +1,146 @@
+namespace firestore {
+namespace util {
+
+// This is a pretty terrible hash implementation for lack of a better one being
+// readily available. It exists as a portability crutch between our existing
+// Objective-C code where overriding `-isEqual:` also requires `-hash` and C++
+// where `operator==()` can be defined without defining a hash code.
+//
+// It's based on the recommendation in Effective Java, Item 9, wherein you
+// implement combosite hashes like so:
+//
+//     size_t result = first_;
+//     result = 31 * result + second_;
+//     result = 31 * result + third_;
+//     return result;
+//
+// This is the basis of this implementation because that's what the existing
+// Objective-C code mostly does by hand. Using this implementation gets the
+// same result by calling
+//
+//     return util::Hash(first_, second_, third_);
+//
+// TODO(wilhuff): Replace this with whatever Abseil releases.
+
+namespace impl {
+
+/**
+ * Combines a hash_value with whatever accumulated state there is so far.
+ */
+inline size_t Combine(size_t state, size_t hash_value) {
+  return 31 * state + hash_value;
+}
+
+/**
+ * Explicit ordering of hashers, allowing SFINAE without all the enable_if
+ * cruft.
+ *
+ * In order we try:
+ *   * A Hash() member, if defined and the return type is an integral type
+ *   * A std::hash specialization, if available
+ *   * A range-based specialization, valid if either of the above hold on the
+ *     members of the range.
+ *
+ * Explicit ordering resolves the ambiguity of the case where a std::hash
+ * specialization is available, but the type is also a range for whose members
+ * std::hash is also available, e.g. with std::string.
+ *
+ * HashChoice is a recursive type that makes each successive specialization more
+ * specialized in the partial ordering among specializations.
+ */
+template <int I>
+struct HashChoice : HashChoice<I + 1> {};
+
+template <>
+struct HashChoice<2> {};
+
+template <typename K>
+size_t InvokeHash(const K& value);
+
+/**
+ * Hashes the given value if it defines a Hash() member.
+ *
+ * @return The result of `value.Hash()`.
+ */
+template <typename K>
+auto InvokeHashHelper(const K& value, HashChoice<0>) -> decltype(value.Hash()) {
+  return value.Hash();
+}
+
+/**
+ * Hashes the given value if it has a specialization of std::hash.
+ *
+ * @return The result of `std::hash<K>{}(value)`
+ */
+template <typename K>
+auto InvokeHashHelper(const K& value, HashChoice<1>)
+    -> decltype(std::hash<K>{}(value)) {
+  return std::hash<K>{}(value);
+}
+
+/**
+ * Hashes the contents of the given range of values if the value_type of the
+ * range can be hashed.
+ */
+template <typename Range>
+auto InvokeHashHelper(const Range& range, HashChoice<2>)
+    -> decltype(impl::InvokeHash(*range.begin())) {
+  size_t result = 0;
+  size_t size = 0;
+  for (auto&& element : range) {
+    size += 1;
+    result = Combine(result, impl::InvokeHash(element));
+  }
+  result = impl::Combine(result, size);
```

Comment:
Question: is hashing size necessary? I think `31` constant makes zero sequences of varying length produce different hashes. Is there a case where size is required to "disambiguate"?

Reply:
Exactly varlength of 0's will all have hash 0 if without length. This appears only an issue for C++ helper here since I believe in objc implementation, all manually implemented hash are actually for fixed-length types.

Reply:
Ah, sorry, I read that it's based on Effective Java algorithm and overlooked that in this case, the algorithm starts from `0` (Effective Java version starts with a non-zero constant, `17`). If you would prefer to follow the hash API outline instead, I'm fine with it.